runtime/compiler/x/codegen/X86PrivateLinkage.cpp

```diff
@@ -2628,6 +2628,107 @@ void J9::X86::PrivateLinkage::buildVPIC(TR::X86CallSite &site, TR::LabelSymbol *
+ *
+ * \parm iterations
+ *    The number of entries on the iTable.
+ *
+ * \parm callNode
+ *    The call node.
+ *
+ * \parm scratchReg
+ *    The temporary register.
+ *
+ * \parm vftReg
+ *    The VFT register.
+ *
+ * \parm interfaceClassReg
+ *    The register that will be used to load the interface class.
+ *
+ * \parm declaringClass
+ *    The class that is to be compared.
+ *
+ * \parm use32BitInterfacePointers
+ *    Whether or not using 32 bit interface class pointer.
+ *
+ * \parm lookupDispatchSnippetLabel
+ *    The label that dispatches to interface look up snippet if the comparison fails.
+ *
+ * \parm gotoLastITableDispatchLabel
+ *    The label that dispatches to invoke the interface call if the comparison succeeds.
+ *
+ * \parm comp
+ *    The compilation object
+ *
+ * \parm cg
+ *    The CodeGenerator object
+ *
+ */
+static void generateITableEntryCompare(uint32_t iterations,
+                                       TR::Node *callNode,
+                                       TR::Register *scratchReg,
+                                       TR::Register *vftReg,
+                                       TR::Register *interfaceClassReg,
+                                       TR_OpaqueClassBlock *declaringClass,
+                                       bool use32BitInterfacePointers,
+                                       TR::LabelSymbol *lookupDispatchSnippetLabel,
+                                       TR::LabelSymbol *gotoLastITableDispatchLabel,
+                                       TR_J9VMBase *fej9,
+                                       TR::Compilation *comp,
+                                       TR::CodeGenerator *cg)
+   {
+   for (uint32_t i = 0; i<iterations; i++)
+      {
+      // First entry
+      if (0 == i)
+         {
+         // scratchReg = j9class->iTable
+         generateRegMemInstruction(TR::InstOpCode::LRegMem(), callNode, scratchReg, generateX86MemoryReference(vftReg, offsetof(J9Class, iTable), cg), cg);
+         }
+      else
+         {
+         // scratchReg = iTable->next
+         generateRegMemInstruction(TR::InstOpCode::LRegMem(), callNode, scratchReg, generateX86MemoryReference(scratchReg, offsetof(J9ITable, next), cg), cg);
+         }
+
+      // Test if iTable == NULL?
+      generateRegRegInstruction(TR::InstOpCode::TESTRegReg(), callNode, scratchReg, scratchReg, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, callNode, lookupDispatchSnippetLabel, cg);
+
+      if (use32BitInterfacePointers)
+         {
+         // The field is 8 bytes, but only 4 matter
+         generateMemImmInstruction(TR::InstOpCode::CMP4MemImm4,
+                                  callNode,
+                                  generateX86MemoryReference(scratchReg, fej9->getOffsetOfInterfaceClassFromITableField(), cg),
+                                  (int32_t)(intptr_t)declaringClass,
```

Comment:
You'll need a relocation for the class pointer immediate if you embed it in the code cache.

Reply:
I'm just guessing here because none of these APIs are documented, but I assume you mean something like this:

```c++
TR::Instruction *cmpInstruction = ...

...

cg->addExternalRelocation(
      new (cg->trHeapMemory()) TR::BeforeBinaryEncodingExternalRelocation(
            cmpInstruction,
            (uint8_t *) (declaringClass),
            TR_ClassPointer,
            cg),
      __FILE__,
      __LINE__,
      callNode);
```